packages/x-data-grid/src/components/columnSelection/GridCellCheckboxRenderer.tsx

```diff
@@ -92,7 +92,11 @@ const GridCellCheckboxForwardRef = React.forwardRef<HTMLInputElement, GridRender
       id,
       rootProps.rowSelectionPropagation?.parents ?? false,
     );
-    const { isIndeterminate, isChecked } = useGridSelector(apiRef, checkboxPropsSelector);
+    const { isIndeterminate, isChecked } = useGridSelector(
+      apiRef,
+      checkboxPropsSelector,
+      objectShallowCompare,
+    );
```

Comment:
The proper solution should be to fix the TODO comment here:

https://github.com/mui/mui-x/blob/90c1fc2e58e53fc1d54219863e8cb6dd6a5356c3/packages/x-data-grid/src/hooks/features/rowSelection/utils.ts#L49-L50

We now have support for selectors with arguments, which is what the checkbox selectors were missing.

Reply:
I might be misunderstanding the new selectors as there's no documentation and only one implementation so far in the code base, but wouldn't it still spit out a new object and trigger a new render, even if the object values stay the same?

Reply:
`createSelectorMemoizedV8` would memoize its output based on its dependencies, e.g.
```js
const checkboxSelector = createSelectorMemoizedV8(
  aSelector,
  bSelector,
  cSelector,
  (a, b, c) => { return { a, b, c } /* this result object would be memoized */ }
)
```

Though I think @MBilalShafi might handle it as part of #15200, so you could ignore this.

Reply:
This part I understand, but wouldn't it only work if a selector returns a slice of the state. In this case we're dealing with derived state, and creating a new object

In this case, you have two options:
* Use a selector that returns the derived state – return value never holds a stable reference, memoization fails.
* Put the state derivation logic in the last function – last function gets triggered also when unrelated rows get selected/deselected. Last function still creates an object with unstable object reference, everything still re-renders when other rows re-render.

So, while it fixes unrelated state triggering re-renders, it still doesn't enable us to hone in on the specific row's state. Either we need to not return an object if the row is not selected, or still need objectShallowCompare to memoize it. 🤔

Reply:
> In this case we're dealing with derived state, and creating a new object

Is it clear to you that the derived state (aka the new object) is memoized by `createSelectorMemoized`? So it stays stable across state updates until one of its dependency (in the example above, `aSelector`, `bSelector` and `cSelector`) changes. That function is similar to `React.useMemo` in its behavior.

Reply:
Ok, I think I understand now after looking at the source code. Args are also passed nested into all of the subselectors, in which case `gridSortedRowIdsSelector` needs to return the correct slice based on the args.

I was first under the impression that args are only passed to the last function that takes together the selector results, in which case it would be triggered by unrelated `gridSortedRowIdsSelector` changes, and we would create a new object every time.

But if args are shared naively across any nested selectors, how do you plan to avoid any conflicts? Seems like something that needs utmost caution if used widely.